velox/type/LongDecimal.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <folly/dynamic.h>
+#include <sstream>
+#include <string>
+#include "velox/common/base/Exceptions.h"
+#include "velox/type/StringView.h"
+
+#pragma once
+
+#define BUILD_INT128(X, Y) ((int128_t)X << 64 | Y)
+
+namespace facebook::velox {
+
+using int128_t = __int128_t;
+
+struct LongDecimal {
+  // Default required for creating vector with NULL values.
+  constexpr explicit LongDecimal() : unscaledValue(0) {}
+  // Constructor to ease creating LongDecimal with literals.
+  constexpr explicit LongDecimal(uint64_t upper, uint64_t lower)
+      : unscaledValue(BUILD_INT128(upper, lower)) {}
+  constexpr explicit LongDecimal(int128_t value) : unscaledValue(value) {}
+
+  bool operator==(const LongDecimal& other) const {
+    return unscaledValue == other.unscaledValue;
+  }
+
+  bool operator!=(const LongDecimal& other) const {
+    return unscaledValue != other.unscaledValue;
+  }
+
+  bool operator<(const LongDecimal& other) const {
+    return unscaledValue < other.unscaledValue;
+  }
+
+  bool operator<=(const LongDecimal& other) const {
+    return unscaledValue <= other.unscaledValue;
+  }
+
+  // Needed for serialization of FlatVector<LongDecimal>
+  operator StringView() const {
+    VELOX_NYI();
+  }
+
+  std::string toString() const;
+
+  operator std::string() const {
+    return toString();
+  }
+
+  operator folly::dynamic() const {
+    VELOX_NYI();
+  }
+
+  int128_t unscaledValue;
+}; // struct LongDecimal
+
+} // namespace facebook::velox
+
+namespace std {
+template <>
+struct hash<::facebook::velox::LongDecimal> {
+  size_t operator()(const ::facebook::velox::LongDecimal& value) const {
+    VELOX_NYI();
+  }
+};
+
+std::string to_string(const ::facebook::velox::LongDecimal& ts);
+
+} // namespace std
+
+namespace folly {
+template <>
+struct hasher<::facebook::velox::LongDecimal> {
+  size_t operator()(const ::facebook::velox::LongDecimal& value) const {
+    VELOX_NYI();
```

Comment:
can we implement this and above using `commutativeHashMix` as in the `inline uint64_t SimpleVector<LongDecimal>::hashValueAt`? Then we do not have to specialize `SimpleVector<LongDecimal>::hashValueAt` at all.